### ISSUE TITLE
[Suggestion] Implement HD Mode 7 for GBA

### ISSUE THREAD

**Issue by frozenLake** [shonumi/gbe-plus#101]
That is, when you get around to increasing the resolution of GBA Games.

Now, before anyone claims this is impossible, do note that [https://i.redd.it/5wphk4fk1is21.png]There was a recent addition to bsnes that does this same thing(url)

Now, this actually can be split into two halves: HD Rotation, and HD Scanline effects. While Rotation will only help the GBA and DS emulation in the future, the Scanline effects may be able to be applied to the gameboy as well, Resulting in certain wavy effects to look far better with CGFX.

Its fine if this request sits around for a long time: I am patient.

**Comment by shonumi** [comment]
It's WIP for now. Some results from 2x sampling:

![sample1](https://i.imgur.com/hYxK1Yk.png)

![sample2](https://i.imgur.com/jPcB32T.png)

Hard to see the differences at 2x, but they're there. 3x and 4x would give the best results. Definitely doable, but very buggy atm. Needs more work.

**Comment by frozenLake** [comment]
So, how does it work with not full mode 7 cases, like the rotation of sprites in Sonic Advance?

**Comment by MojoJojoDojo** [comment]
Does this apply to any DS game that uses 2D effects like scaling and rotation?
Or it needs to be done separably.

**Comment by shonumi** [comment]
The DS' affine transformations are copy+paste from the GBA, so yes, it applies to the NDS as much as the GBA. However, most DS games (if any?) did the psuedo Mode-7 like MK:SS or F-Zero. They still did it for sprites and backgrounds.

**Comment by MojoJojoDojo** [comment]
Chrono Trigger DS and possibly Kirby Super Star Ultra used them as they are ports from the SNES.
CT DS used the feature for sure because i checked upscaling with Desmume and it didnt upscale them which means they use the 2D engine for "Mode-7".

**Comment by shonumi** [comment]
Yeah, good catch. You'd probably only find the psuedo Mode 7 in a few other SNES ports. Most 2D games (original NDS titles) just scale/rotate sprites and a couple of BG layers from what I've seen though. The effect certainly got down played after the GBA.
